src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -112,34 +115,110 @@ private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory)
+
+            if (Process.GetCurrentProcess().ProcessName.Equals("dotnet"))
+            {
+                // We're already in a dotnet process! Try using it.
+                dotnetPath = Path.GetDirectoryName(Process.GetCurrentProcess().MainModule.FileName);
+                foreach (string path in GetDotNetBasePaths(workingDirectory, dotnetPath))
+                {
+                    returnedSDKs = true;
+                    if (pinnedSdk)
+                    {
+                        pinnedSdk = false;
+                        pathOfSecondPinnedSdk = path;
+                    }
+                    else
+                    {
+                        sdkPaths.Add(path);
+                    }
+                }
+            }
+
+            if (!returnedSDKs && rc != 0)
+            {
+                throw new InvalidOperationException("Failed to find an appropriate version of .NET Core MSBuild. Call to hostfxr_resolve_sdk2 failed. There may be more details in stderr.");
+            }
+            else if (returnedSDKs)
+            {
+                List<string> discoveredSDKs = sdkPaths.ToList();
+                discoveredSDKs.Sort();
+
+                if (!string.IsNullOrEmpty(pathOfPinnedSdk))
+                {
+                    discoveredSDKs.Add(pathOfPinnedSdk);
+                }
+
+                if (!string.IsNullOrEmpty(pathOfSecondPinnedSdk))
+                {
+                    discoveredSDKs.Add(pathOfSecondPinnedSdk);
+                }
+
+                // We currently have the SDKs sorted in increasing version number with the pinned SDKs at the end. Reverse that order.
+                discoveredSDKs.Reverse();
+
+                return discoveredSDKs;
+            }
+            else
+            {
+                return Enumerable.Empty<string>();
+            }
+        }
+
+        private static IEnumerable<string> GetDotNetBasePaths(string workingDirectory, string dotnetPath)
+        {
             string bestSDK = null;
-            int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
+            
+            // Only overwrite the return code if we would set it to a non-zero value so that if we ultimately fail to find any SDKs, we can provide
+            // a reasonable error. Otherwise, we would only ever see failures from the last call to GetDotNetBasePaths.
+            int localRC = 0;
+            localRC = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
             {
                 if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
                 {
                     bestSDK = value;
                 }
             });
 
-            if (rc == 0 && bestSDK != null)
+            if (localRC == 0 && bestSDK != null)
             {
+                pinnedSdk = true;
```

Comment:
What is a "pinned" sdk?

Reply:
That would be like if you have a global.json. Is there a better name that would've made that clear?